cmd/sql-main.go

```diff
@@ -68,11 +101,279 @@ EXAMPLES:
+		fatalIf(err, "Invalid serialization option(s) specified for --csv-input flag")
+
+		m["csv"] = kv
+	}
+	if ijson != "" {
+		kv, err := parseSerializationOpts(ijson, validJSONInputKeys, nil)
+
+		fatalIf(err, "Invalid serialization option(s) specified for --json-input flag")
+		m["json"] = kv
+	}
+
+	return m
+}
+
+// gets the output serialization opts from cli context and constructs a map of csv or json options
+func getOutputSerializationOpts(ctx *cli.Context, csvHdrs []string) (opts map[string]map[string]string) {
+	m := make(map[string]map[string]string)
+	var csvType, jsonType bool
+
+	ocsv := ctx.String("csv-output")
+	ojson := ctx.String("json-output")
+	csvType = (ocsv != "")
+	jsonType = (ojson != "")
+
+	if csvType && jsonType {
+		fatalIf(errInvalidArgument(), "Only one of --csv-output, or --json-output can be specified as output serialization option")
+	}
+
+	if jsonType && len(csvHdrs) > 0 {
+		fatalIf(errInvalidArgument(), "--csv-output-header incompatible with --json-output option")
+	}
+
+	if ocsv != "" {
+		validKeys := append(validCSVCommonKeys, validJSONCSVCommonOutputKeys...)
+		kv, err := parseSerializationOpts(ocsv, append(validKeys, validCSVOutputKeys...), validCSVOutputAbbrKeys)
+		fatalIf(err, "Invalid value(s) specified for --csv-output flag")
+		m["csv"] = kv
+	}
+
+	if ojson != "" {
+		kv, err := parseSerializationOpts(ojson, validJSONCSVCommonOutputKeys, validJSONOutputAbbrKeys)
+		fatalIf(err, "Invalid value(s) specified for --json-output flag")
+		m["json"] = kv
+	}
+	return m
+}
+
+// getCSVHeader fetches the first line of csv query object
+func getCSVHeader(sourceURL string, encKeyDB map[string][]prefixSSEPair) ([]string, *probe.Error) {
+	var r io.ReadCloser
+	switch sourceURL {
+	case "-":
+		r = os.Stdin
+	default:
+		var err *probe.Error
+		var metadata map[string]string
+		if r, metadata, err = getSourceStreamMetadataFromURL(sourceURL, encKeyDB); err != nil {
+			return nil, err.Trace(sourceURL)
+		}
+		ctype := metadata["Content-Type"]
+		if strings.Contains(ctype, "gzip") {
+			var e error
+			r, e = gzip.NewReader(r)
+			if e != nil {
+				return nil, probe.NewError(e)
+			}
+			defer r.Close()
+		} else if strings.Contains(ctype, "bzip") {
+			defer r.Close()
+			r = ioutil.NopCloser(bzip2.NewReader(r))
+		} else {
+			defer r.Close()
+		}
+	}
+	br := bufio.NewReader(r)
+	line, _, err := br.ReadLine()
+	if err != nil {
+		return nil, probe.NewError(err)
+	}
+	return strings.Split(string(line), ","), nil
+}
+
+// if csv-output-header is set to a comma delimited string use it, othjerwise attempt to get the header from
+// query object
+func getCSVOutputHeaders(ctx *cli.Context, url string, encKeyDB map[string][]prefixSSEPair, query string) (hdrs []string) {
+	if !ctx.IsSet("csv-output-header") {
+		return
+	}
+
+	hdrStr := ctx.String("csv-output-header")
+	if hdrStr == "" && strings.HasPrefix(strings.ToLower(query), "select *") {
```

Comment:
Why is the second check needed here?

Reply:
@donatello, csv header is being displayed if query is selecting all columns of the csv object. Hence the second check.

Reply:
Oh ok, maybe the check can be put in its own function and also be less strict about whitespace - maybe a regexp like `\s+select\s+\*\s+` is better.

Reply:
done